File: src/python/command_injection.py
"""CWE-78: OS Command Injection via os.system / subprocess shell=True."""

import os
import subprocess
from flask import Flask, request

app = Flask(__name__)


@app.route("/ping")
def ping():
    host = request.args.get("host", "")
    return os.popen("ping -c 1 " + host).read()


@app.route("/dns")
def dns():
    domain = request.args.get("domain", "")
    return subprocess.check_output(f"dig {domain}", shell=True).decode()


@app.route("/archive")
def archive():
    path = request.args.get("path", "")
    os.system(f"tar -czf /tmp/out.tgz {path}")
    return "ok"


if __name__ == "__main__":
    app.run(port=4002)

File: src/javascript/command-injection.js
// CWE-78: OS Command Injection via child_process.

const express = require('express');
const { exec, execSync } = require('child_process');

const app = express();

app.get('/ping', (req, res) => {
  const host = req.query.host;
  exec('ping -c 1 ' + host, (err, stdout, stderr) => {
    if (err) return res.status(500).send(stderr);
    res.type('text/plain').send(stdout);
  });
});

app.get('/dns', (req, res) => {
  const domain = req.query.domain;
  const out = execSync(`dig ${domain}`).toString();
  res.type('text/plain').send(out);
});

app.get('/backup', (req, res) => {
  const path = req.query.path;
  exec(`tar -czf /tmp/backup.tgz ${path}`, (err) => {
    if (err) return res.status(500).send(err.message);
    res.send('ok');
  });
});

app.listen(3002);
